### PR TITLE
Feature/tray icon dbus sni test (merge to #6560)

### PR DESCRIPTION
**Pull request by jmacato**
cc @kekekeks @danwalmsley 

targets: #6560

### CODE REVIEW

**Comment by jmacato** [comment]
@kekekeks im having trouble linking the dbus menu exporter to the SNI object.. idk how to move forward on this one

File: src/Avalonia.FreeDesktop/DBusSystemTray/DBusSysTray.cs [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+using Avalonia.Controls.Platform;
+using Tmds.DBus;
+
+[assembly: InternalsVisibleTo(Tmds.DBus.Connection.DynamicAssemblyName)]
+
+namespace Avalonia.FreeDesktop.DBusSystemTray
+{
+    public class DBusSysTray : IDisposable
```
Comment (danwalmsley) [inline src/Avalonia.FreeDesktop/DBusSystemTray/DBusSysTray.cs:14]: Seems this class can be the one that implements ITrayIconImpl?

File: src/Avalonia.X11/X11TrayIconImpl.cs [hunk 2]
```diff
@@ -1,33 +1,88 @@
...
         public void SetIsVisible(bool visible)
         {
-            
+            if (visible && lastIcon != null)
+            {
+                SetIcon(lastIcon);
+            }
+            else
+            {
+                _dBusSysTray.SetIcon(new DbusPixmap(1, 1, new byte[] { 0, 0, 0, 0 }));
```
Comment (danwalmsley) [inline src/Avalonia.X11/X11TrayIconImpl.cs:79]: is there not a way to actually remove the icon completely. This is the api to remove the icon all together.

**Review by danwalmsley — CHANGES_REQUESTED** [review]

**Review by ghost — CHANGES_REQUESTED** [review]
tested a7f3fbe on Qubes 4 Debian 10 i3wm

it crashes with
```
Tmds.DBus.DBusException: org.freedesktop.DBus.Error.ServiceUnknown: The name com.canonical.AppMenu.Registrar was not provided by any .service files
   at Tmds.DBus.DBusConnection.CallMethodAsync(Message msg, Boolean checkConnected, Boolean checkReplyType)
   at Tmds.DBus.Connection.CallMethodAsync(Message message)
   at Tmds.DBus.CodeGen.DBusObjectProxy.SendMethodReturnReaderAsync(String iface, String member, Nullable`1 inSignature, MessageWriter writer)
   at Avalonia.FreeDesktop.DBusMenuExporter.DBusMenuExporterImpl.Init() in /home/user/Avalonia/src/Avalonia.FreeDesktop/DBusMenuExporter.cs:line 79
Unhandled exception. Tmds.DBus.DBusException: org.freedesktop.DBus.Error.ServiceUnknown: The name org.kde.StatusNotifierWatcher was not provided by any .service files
   at Tmds.DBus.DBusConnection.CallMethodAsync(Message msg, Boolean checkConnected, Boolean checkReplyType)
   at Tmds.DBus.Connection.CallMethodAsync(Message message)
   at Tmds.DBus.CodeGen.DBusObjectProxy.SendMethodReturnReaderAsync(String iface, String member, Nullable`1 inSignature, MessageWriter writer)
   at Avalonia.X11.X11TrayIconImpl.CreateTrayIcon() in /home/user/Avalonia/src/Avalonia.X11/X11TrayIconImpl.cs:line 64
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__139_0(Object state)
   at Avalonia.Threading.AvaloniaSynchronizationContext.<>c__DisplayClass5_0.<Post>b__0() in /home/user/Avalonia/src/Avalonia.Base/Threading/AvaloniaSynchronizationContext.cs:line 33
   at Avalonia.Threading.JobRunner.Job.Avalonia.Threading.JobRunner.IJob.Run() in /home/user/Avalonia/src/Avalonia.Base/Threading/JobRunner.cs:line 166
   at Avalonia.Threading.JobRunner.RunJobs(Nullable`1 priority) in /home/user/Avalonia/src/Avalonia.Base/Threading/JobRunner.cs:line 37
   at Avalonia.Threading.Dispatcher.EnsurePriority(DispatcherPriority currentPriority) in /home/user/Avalonia/src/Avalonia.Base/Threading/Dispatcher.cs:line 123
   at Avalonia.Threading.DispatcherTimer.InternalTick() in /home/user/Avalonia/src/Avalonia.Base/Threading/DispatcherTimer.cs:line 209
   at Avalonia.X11.X11PlatformThreading.RunLoop(CancellationToken cancellationToken) in /home/user/Avalonia/src/Avalonia.X11/X11PlatformThreading.cs:line 225
   at Avalonia.Threading.Dispatcher.MainLoop(CancellationToken cancellationToken) in /home/user/Avalonia/src/Avalonia.Base/Threading/Dispatcher.cs:line 61
   at Avalonia.Controls.ApplicationLifetimes.ClassicDesktopStyleApplicationLifetime.Start(String[] args) in /home/user/Avalonia/src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs:line 126
   at Avalonia.ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime[T](T builder, String[] args, ShutdownMode shutdownMode) in /home/user/Avalonia/src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs:line 175
   at ControlCatalog.NetCore.Program.Main(String[] args) in /home/user/Avalonia/samples/ControlCatalog.NetCore/Program.cs:line 100
```

**Comment by jmacato** [comment]
@MaxHillebrand we will probably need XEmbed fallback for linux OS's without the SNI support

**Comment by MaxHillebrand** [comment]
after building and executing https://git.sr.ht/~steef/snixembed
it does work flawlessly!!
screenshot_20210924_105108

Amazing work @jmacato :black_heart:

**Pull request merged**